### PR TITLE
Do not use stale RegistrationSettings if the User Registration feature is disabled.

### PR DESCRIPTION
**Pull request by mvarblow**
Do not use stale RegistrationSettings if the User Registration feature is disabled.

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 1]
```diff
@@ -958,6 +950,29 @@ private async Task<bool> AddConfirmEmailErrorAsync(IUser user)
         return false;
     }
 
+    private async Task<RegistrationSettings> GetRegistrationSettingsIfEnabledAsync()
+    {
+        var registrationFeatureIsAvailable = (await _shellFeaturesManager.GetEnabledFeaturesAsync())
+            .Any(feature => feature.Id == UserConstants.Features.UserRegistration);
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:956]:
```suggestion
        var isRegistrationFeatureEnabled = await _shellFeaturesManager.IsFeatureEnabledAsync(UserConstants.Features.UserRegistration);
```

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 2]
```diff
@@ -958,6 +950,29 @@ private async Task<bool> AddConfirmEmailErrorAsync(IUser user)
...
+
+        if (!registrationFeatureIsAvailable)
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:958]:
```suggestion
        if (!isRegistrationFeatureEnabled)
```

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 3]
```diff
@@ -958,6 +950,29 @@ private async Task<bool> AddConfirmEmailErrorAsync(IUser user)
...
+            return null;
+        }
+
+        return await _siteService.GetSettingsAsync<RegistrationSettings>();
+    }
+
+    private static bool UsersMustValidateEmail(RegistrationSettings registrationSettings)
+    {
+        return registrationSettings?.UsersMustValidateEmail ?? false;
+    }
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:969]:
```suggestion
    private static bool UsersMustValidateEmail(RegistrationSettings registrationSettings) => registrationSettings?.UsersMustValidateEmail ?? false;
```

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 4]
```diff
@@ -958,6 +950,29 @@ private async Task<bool> AddConfirmEmailErrorAsync(IUser user)
...
+
+    private static bool IsUserRegistrationDisabled(RegistrationSettings registrationSettings)
+    {
+        return registrationSettings == null || registrationSettings.UsersCanRegister == UserRegistrationType.NoRegistration;
+    }
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:974]:
```suggestion
    private static bool IsUserRegistrationDisabled(RegistrationSettings registrationSettings)
        => registrationSettings == null || registrationSettings.UsersCanRegister == UserRegistrationType.NoRegistration;
```

**Review by hishamco — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 5]
```diff
@@ -958,6 +950,23 @@ private async Task<bool> AddConfirmEmailErrorAsync(IUser user)
...
+
+        if (!isRegistrationFeatureEnabled)
+        {
+            return null;
+        }
+
+        return await _siteService.GetSettingsAsync<RegistrationSettings>();
+    }
+
+    private static bool UsersMustValidateEmail(RegistrationSettings registrationSettings) => registrationSettings?.UsersMustValidateEmail ?? false;
```
[the remark this answers is not in the document]
Comment (mvarblow) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:965]: Your suggested method name seems wrong to me. This method doesn't tell you if the user email is validated. It tells you if the registration settings require that the user's email be validated.

[the remark this answers is not in the document]
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:965]: What about now?

[the remark this answers is not in the document]
Comment (mvarblow) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:965]: That's fine, but I'm not sure how it's better than UsersMustValidateEmail. :)

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 6]
```diff
@@ -958,6 +950,23 @@ private async Task<bool> AddConfirmEmailErrorAsync(IUser user)
         return false;
     }
 
+    private async Task<RegistrationSettings> GetRegistrationSettingsIfEnabledAsync()
```
[the remark this answers is not in the document]
Comment (mvarblow) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:953]: I almost called it GetRegistrationSettingsOrNullAsync. Seems like it's better for the name to make clear what it's doing beyond just getting the registration settings?

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:965]: The method always starts with verbs

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:953]: Much better `GetRegistrationSettingsOrDefaultAsync`

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 7]
```diff
@@ -958,6 +950,23 @@ private async Task<bool> AddConfirmEmailErrorAsync(IUser user)
...
+    {
+        var isRegistrationFeatureEnabled = await _shellFeaturesManager.IsFeatureEnabledAsync(UserConstants.Features.UserRegistration);
+
+        if (!isRegistrationFeatureEnabled)
+        {
+            return null;
+        }
+
+        return await _siteService.GetSettingsAsync<RegistrationSettings>();
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:962]:
```suggestion
        return isRegistrationFeatureEnabled ? await _siteService.GetSettingsAsync<RegistrationSettings>() : null;
```

**Review by hishamco — CHANGES_REQUESTED** [review]

File: src/Templates/OrchardCore.ProjectTemplates/content/OrchardCore.Templates.Mvc.Module/.template.config/dotnetcli.host.json [hunk 8]
```diff
@@ -12,6 +12,6 @@
   },
   "usageExamples": [
     "--framework net8.0",
-    "--orchard-version 2.0.0"
+    "--orchard-version 1.8.3"
```
Comment (hishamco) [inline src/Templates/OrchardCore.ProjectTemplates/content/OrchardCore.Templates.Mvc.Module/.template.config/dotnetcli.host.json:15]: These changes are not related to this PR, right?

Comment (mvarblow) [reply src/Templates/OrchardCore.ProjectTemplates/content/OrchardCore.Templates.Mvc.Module/.template.config/dotnetcli.host.json]: No. Sorry. That was not intentional! I've reverted those changes.

**Review by hishamco — APPROVED** [review]

**Comment by hishamco** [comment]
@MikeAlhayek @gvkries Does any one have an objection or shall we merge this?

**Comment by MikeAlhayek** [comment]
@hishamco this PR may fix the problem. But I think the better solution is PR #16721 which extracts all the external user logic and settings into a separate feature that is disabled by default. Yes there is a fast-forward path that would auto enable it if the user is using external authentication.

I'll try to finalize it tomorrow. That PR also transform the settings into more logical places so that we can have better control on the settings.

In short, wait on this PR until I am done with the other PR just to see if there is still a reason to take this PR. If you take this PR, my PR will override it all.

**Comment by hishamco** [comment]
IMHO let us take this while @mvarblow started this first unless your PR is much much better :)

I will hold this PR, thanks again @mvarblow

**Comment by Piedone** [comment]
Yeah, https://github.com/OrchardCMS/OrchardCore/pull/16721 seems like a better approach.
